gateway-ha/src/main/java/io/trino/gateway/ha/router/QueryCountBasedRouter.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.trino.gateway.ha.router;
+
+import io.trino.gateway.ha.clustermonitor.ClusterStats;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
+
+public class QueryCountBasedRouter
+            extends HaRoutingManager
+{
+    private ArrayList<ClusterStats> clusterStats;
+    private static final ReentrantReadWriteLock lock = new ReentrantReadWriteLock(true);
+
+    public QueryCountBasedRouter(GatewayBackendManager gatewayBackendManager,
+                                 QueryHistoryManager queryHistoryManager)
+    {
+        super(gatewayBackendManager, queryHistoryManager);
+        clusterStats = new ArrayList<>();
+    }
+
+    // We sort and find the backend based on the individual user's count of the queued queries
+    // first, in case user doesn't have any queries queued we use the cluster wide stats
+    //
+    // First filter the list of clusters for a particular routing group
+    //
+    // If a user has queued queries, then find a cluster with the least number of QUEUED queries
+    // for that user.
+    //
+    // If a user's QUEUED query count is the same on every cluster then go with a cluster with
+    // cluster wide stats.
+    //
+    // Find a cluster with the least number of QUEUED queries, if there are the same number of
+    // queries queued, then compare the number of running queries.
+    //
+    // After a query is routed, we need to update the stats for that cluster until we received the
+    // updated stats for all the clusters.
+    // if a user has queries queued then we assume that the routed query will be also queued or
+    // else we assume it would be scheduled immediately and we increment the stats for the running
+    // queries
+
+    private Optional<ClusterStats> getClusterToRoute(String user, String routingGroup)
+    {
+        List<ClusterStats> filteredList = null;
+        Lock readLock = lock.readLock();
+        try {
+            readLock.lock();
+            filteredList = clusterStats.stream()
```

Comment:
This would be a good use case for a `Map<Array<ClusterStats>>` or potentially `com.google.common.collect.Table` if you need to look up on various attributes vs filtering a list

Reply:
Yeah, If it becomes more complicated then yes, may be the next iteration when we get more stats for the cost based routing we can use it. At this time the filter and sort is just simple and clean IMO.